Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc

```diff
@@ -55,13 +57,26 @@ void ExponentialBackoff::BackoffAndRun(AsyncQueue::Operation&& operation) {
 
   // First schedule the block using the current base (which may be 0 and should
   // be honored as such).
-  Milliseconds delay_with_jitter = current_base_ + GetDelayWithJitter();
-  if (delay_with_jitter.count() > 0) {
-    LOG_DEBUG("Backing off for %s milliseconds (base delay: %s milliseconds)",
-              delay_with_jitter.count(), current_base_.count());
+  Milliseconds desired_delay_with_jitter = current_base_ + GetDelayWithJitter();
+
+  Milliseconds delay_so_far = chr::duration_cast<Milliseconds>(
+      chr::steady_clock::now() - last_attempt_time_);
+
+  // Guard against the backoff delay already being past.
+  auto remaining_delay =
+      max(Milliseconds::zero(), desired_delay_with_jitter - delay_so_far);
+
+  if (current_base_.count() > 0) {
+    LOG_DEBUG(
+        "Backing off for %s ms "
+        "(base delay: %s ms, "
+        "delay with jitter: %s ms, "
+        "last attempt: %s ms ago)",
+        remaining_delay.count(), current_base_.count(),
+        desired_delay_with_jitter.count(), delay_so_far.count());
   }
 
-  delayed_operation_ = queue_->EnqueueAfterDelay(delay_with_jitter, timer_id_,
+  delayed_operation_ = queue_->EnqueueAfterDelay(remaining_delay, timer_id_,
                                                  std::move(operation));
```

Comment:
Hmm, shouldn't `last_update_time_` be reset here? You can wrap the `operation` into another `std::function`, for example:
```cpp
delayed_operation_ = queue_->EnqueueAfterDelay(remaining_delay, timer_id_,
// I'm using a lambda here; the actual parameter is `std::function<void()>`, and
// `std::function` can be implicitly initialized by a lambda with corresponding signature.
  [this, operation] {
    last_attempt_time_ = chr::steady_clock::now();
    operation();
});
```
The only downside is having to copy `operation` into the lambda (this `[operation]` syntax of the capture performs a copy). It would be ideal to move, but unfortunately it's only possible starting from C++14, while we have to support C++11. These unnecessary copies already happen in `AsyncQueue`, so I think it's not really a big deal.

Update: I see the Objective-C version also does the strong/weak self thing. I think in C++ streams should always cancel the backoff on shutdown, so lifetime shouldn't be an issue.